app/src/main/java/com/gusta/wakemehome/geofencing/GeofenceManager.java

```diff
@@ -0,0 +1,375 @@
+            mPendingGeofenceTask = PendingGeofenceTask.REFRESH;
+            requestPermissions();
+            return;
+        }
+
+        // TODO: remove only irrelevant geofences and add only the new ones
+        // Remove all geofences and back all the relevant ones
+        removeGeofences();
+        addGeofences();
+    }
+
+    /**
+     * Builds and returns a GeofencingRequest. Specifies the list of geofences to be monitored.
+     * Also specifies how the geofence notifications are initially triggered.
+     */
+    private GeofencingRequest getGeofencingRequest() {
+        Log.d(TAG, "Creating and destroying geofences according to alarms list");
+        List<? extends GeofenceEntry> entries = mLiveData.getValue();
+        List<Geofence> geofenceList = new ArrayList<>();
+
+        // Create geofence objects
+        if (entries != null) {
+            for(GeofenceEntry entry : entries)
+            {
+                if (entry.isEnabled()) {
+                    geofenceList.add(new Geofence.Builder()
+                            // Set the request ID of the geofence. This is a string to identify this
+                            // geofence.
+                            .setRequestId(String.valueOf(entry.getId()))
+
+                            .setCircularRegion(
+                                    entry.getLatitude(),
+                                    entry.getLongitude(),
+                                    entry.getRadius()
+                            )
+                            .setExpirationDuration(Geofence.NEVER_EXPIRE)
+                            .setTransitionTypes(Geofence.GEOFENCE_TRANSITION_ENTER)
+                            .build());
+                }
+            }
+        }
+
+        // Specify geofences and initial triggers
+        GeofencingRequest.Builder builder = new GeofencingRequest.Builder();
+        builder.setInitialTrigger(GeofencingRequest.INITIAL_TRIGGER_ENTER);
+        builder.addGeofences(geofenceList);
+        return builder.build();
+    }
+
+    /**
+     * Adds geofences. This method should be called after the user has granted the location
+     * permission.
+     */
+    private void addGeofences() {
+        if (!checkPermissions()) {
+            showSnackbar(mActivity.getString(R.string.insufficient_permissions));
+            return;
+        }
+        mGeofencingClient.addGeofences(getGeofencingRequest(), getGeofencePendingIntent())
+                .addOnSuccessListener(mActivity, new OnSuccessListener<Void>() {
+                    @Override
+                    public void onSuccess(Void aVoid) {
+                        Log.d(TAG, "Geofences added");
+                    }
+                })
+                .addOnFailureListener(mActivity, new OnFailureListener() {
+                    @Override
+                    public void onFailure(@NonNull Exception e) {
+                        //TODO: Handle StatusCode = 1000: "Geofence service is not available now.
```

Comment:
No needed retry?

Reply:
I think not, you could say no need second retry later. The right way is a job instead of any constant number of retryes - but this is not core functionality as long as the user is told that the alarms are not set (as added in next brunch)